[PATCH v4] testcases: fix file path to control access to cron

lucasrangit opened the pull request:
crontab uses /etc/cron.{allow,deny} to control access on most distributions, as well as RHEL. To maintain backwards compatibility with older still supported distributions, the test setup will search the man page for the correct paths (e.g. /var/spool/cron/allow). If the man page is not available (e.g. on an embedded device), it will default to the /etc paths. This method supports vixie-cron, cronie, etc...

The repeating code has been moved into a common shell script library to simplify maintenance.

Reported-by: Myungho Jung <mhjungk@gmail.com>
Signed-off-by: Lucas Magasweran <lucas.magasweran@ieee.org>
Tested-by: Dan Rue <drue@therub.org>

---

Changes in v2: 
* This based on the feedback in https://github.com/linux-test-project/ltp/pull/149 . I could not figure out a reliable way to detect the path based on the distribution since each distribution uses a different cron (e.g. cronie, vixie cron, etc.) and this varies across distro releases. I even tried `strace` to see which file `crontab -l` opened. This was reliable but may break in the future.

Changes in v3:
* Require a leading slash to match absolute paths in RHEL7.4.
* Require trailing whitespace to the end of line to not match multiple occurrences with cron 3.0pl1-128ubuntu2 man page on Ubuntu 16.04.

Changes in v4:
* Simplify pattern matching to expect the end of line after the path in the FILES section.

metan-ucw commented on testcases/commands/cron/cron_allow01:
```diff
@@ -31,11 +31,16 @@ iam=`whoami`
+	echo "Error while searching for cron access files"
+	exit 1
 fi
 
```
Can we, pretty please, put this repeating snippet of code in a cron_common.sh file and source it in all the tests?

It would be much easier to adjust the logic that way.

lucasrangit replied:
I agree!

Would you like two new commits; the first to create the `cron_common.sh` and second to change the logic? Or can I simply make a new commit on top of this to refactor the code into the common file?

metan-ucw replied:
Just amend the change to this commit so that we end up just with one commit that both adds the library for the common code and makes the tests to use that.

lucasrangit commented:
@metan-ucw , as requested I created the `cron_common.sh` and moved the repeating code.

After some more testing, I found that I could use the `crontab` man page to determine the path to the `{allow,deny}` files. This works all the way back to SuSE Linux 8.2 and RHEL 4. Hope you like it.

lucasrangit commented:
Thanks for testing this patch @danrue .

jstancek commented:
My man page on RHEL7.4 has multiple strings that match, and path to cron.deny is not the first one:
```
$ man crontab | grep -m 1 -o '[\/a-z.]*deny\b'
cron.deny
```
but this could be tweaked in some way:
```
man crontab | grep -m 1 -o '/[\/a-z.]*deny\b'
/etc/cron.deny
```

One corner case could be small systems that don't have man pages installed, but our options seems quite limited.

My only other idea is:
```
strings $(which crontab) | grep '/cron.allow'
```

lucasrangit commented:
@jstancek Thanks for testing this on RHEL7.4. I reproduced the issue with the regex and will try and fix it. Thanks for your suggested change.

> ```
> strings $(which crontab) | grep '/cron.allow'
> ```

I thought about this as well since it should provide the string *used* in the binary. But on some systems the file has a completely different name, such as `/var/spool/cron/allow`, which I found harder to grep for.

> One corner case could be small systems that don't have man pages installed, but our options seems quite limited.

Very true. Are those embedded systems officially supported by LTP? Also, if they don't include `man` do they include the `strings`? I've only used `strings` to debug.

danrue commented:
I use LTP on both ARM based server systems with full operating systems, as well as ARM based embedded boards running openembedded and android.

I checked on an embedded board running openembedded, and the strings method worked with gnu strings as well as busybox strings. The man page method would not work, because man pages are not installed.

As far as whether or not LTP is used to test embedded systems: yes, absolutely. But I don't run all the tests. admin_tools is not run because it is assuming a full user space. For example, su is not installed so there is no point in testing it.

lucasrangit commented:
> I checked on an embedded board running openembedded, and the strings method worked with gnu strings as well as busybox strings. The man page method would not work, because man pages are not installed.

@danrue, does the `export CRON_DENY="$(man crontab 2>/dev/null | grep -m 1 -o '[\/a-z.]*deny\b' || echo "/etc/cron.deny")"` properly fail and use the "/etc/" path on your embedded system?

danrue commented:
Yes the /etc/cron.deny default works fine.

```sh
root@hikey:~# export CRON_DENY="$(man crontab 2>/dev/null | grep -m 1 -o '[\/a-z.]*deny\b' || echo "/etc/cron.deny")"
root@hikey:~# echo $CRON_DENY
/etc/cron.deny
root@hikey:~# strings $(which crontab) | grep '/cron.deny'
/etc/cron.deny
root@hikey:~# busybox strings $(which crontab) | grep '/cron.deny'
/etc/cron.deny
```

lucasrangit commented:
@jstancek , I have incorporated your change to require a leading slash. I also found that the Ubuntu 16.04 (vixie-cron) cron 3.0pl1-128ubuntu2 man page matches `/etc/cron.deny` multiple times on a single line. Since `grep` doesn't handle this without another pipe to `head -1`, which would break the fall back to the `/etc` paths (without `set -o pipefail`), I added a regex match for trailing whitespace to the end of line.

```
export CRON_DENY="$(man crontab 2>/dev/null | grep -m 1 -o '/[\/a-z.]*deny\b[[:space:]]*$' || echo "/etc/cron.deny")"
export CRON_ALLOW="$(man crontab 2>/dev/null | grep -m 1 -o '/[\/a-z.]*allow\b[[:space:]]*$' || echo "/etc/cron.allow")"
```

What do you think?

lucasrangit commented:
Re-tested on Ubuntu 16.04.3 with the latest rebased on top of master.

```
$ sudo /opt/ltp/runltp -f admin_tools -s cron
INFO: creating /opt/ltp/output directory
INFO: creating /opt/ltp/results directory
Checking for required user/group ids

'nobody' user id and group found.
'bin' user id and group found.
'daemon' user id and group found.
Users group found.
Sys group found.
Required users/groups exist.
If some fields are empty or look unusual you may have an old version.
Compare to the current minimal requirements in Documentation/Changes.
 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.3 LTS"
NAME="Ubuntu"
VERSION="16.04.3 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.3 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
Linux ubuntu1604.localdomain 4.4.0-97-generic #120-Ubuntu SMP Tue Sep 19 17:28:18 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
 
Gnu C                  gcc (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
Gnu make               4.1
util-linux             linux 2.27.1
mount                  linux 2.27.1 (libmount 2.27.0: selinux, assert, debug)
modutils               22
e2fsprogs              1.42.13
Linux C Library        > libc.2.23
Dynamic linker (ldd)   2.23
Procps                 3.3.10
Net-tools              1.60
iproute2               iproute2-ss151103
iputils                iputils-s20121221
Kbd                    1.15.5
Sh-utils               8.25
Modules Loaded         vmw_vsock_vmci_transport vsock ppdev vmw_balloon input_leds coretemp joydev serio_raw i2c_piix4 shpchp nfit vmw_vmci 8250_fintek parport_pc parport mac_hid ib_iser rdma_cm iw_cm ib_cm ib_sa ib_mad ib_core ib_addr iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsi autofs4 btrfs raid10 raid456 async_raid6_recov async_memcpy async_pq async_xor async_tx xor raid6_pq libcrc32c raid1 raid0 multipath linear crct10dif_pclmul crc32_pclmul ghash_clmulni_intel vmwgfx ttm drm_kms_helper aesni_intel syscopyarea sysfillrect sysimgblt aes_x86_64 lrw gf128mul glue_helper fb_sys_fops ablk_helper cryptd psmouse floppy e1000 mptspi mptscsih mptbase drm scsi_transport_spi pata_acpi fjes

free reports:
              total        used        free      shared  buff/cache   available
Mem:        2048176       77544      300800       10452     1669832     1779952
Swap:       4193276           0     4193276

/proc/cpuinfo
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 94
model name	: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
stepping	: 3
microcode	: 0xba
cpu MHz		: 2712.000
cache size	: 8192 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 22
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm 3dnowprefetch epb fsgsbase smep dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bugs		:
bogomips	: 5424.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:

processor	: 1
vendor_id	: GenuineIntel
cpu family	: 6
model		: 94
model name	: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
stepping	: 3
microcode	: 0xba
cpu MHz		: 2712.000
cache size	: 8192 KB
physical id	: 2
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 2
initial apicid	: 2
fpu		: yes
fpu_exception	: yes
cpuid level	: 22
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm 3dnowprefetch epb fsgsbase smep dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bugs		:
bogomips	: 5424.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:

no big block device was specified on commandline.
Tests which require a big block device are disabled.
You can specify it with option -z
COMMAND:    /opt/ltp/bin/ltp-pan   -e -S   -a 16037     -n 16037 -p -f /tmp/ltp-B5ZYE0aUcX/alltests -l /opt/ltp/results/LTP_RUN_ON-2017_11_02-16h_35m_21s.log  -C /opt/ltp/output/LTP_RUN_ON-2017_11_02-16h_35m_21s.failed -T /opt/ltp/output/LTP_RUN_ON-2017_11_02-16h_35m_21s.tconf
INFO: Restricted to cron
LOG File: /opt/ltp/results/LTP_RUN_ON-2017_11_02-16h_35m_21s.log
FAILED COMMAND File: /opt/ltp/output/LTP_RUN_ON-2017_11_02-16h_35m_21s.failed
TCONF COMMAND File: /opt/ltp/output/LTP_RUN_ON-2017_11_02-16h_35m_21s.tconf
Running tests.......
<<<test_start>>>
tag=cron02 stime=1509665722
cmdline="cron02"
contacts=""
analysis=exit
<<<test_output>>>
userdel: user 'c01_user' does not exist
Running as user c01_user...
new job added successfully
sleeping for 130 seconds...
Testjob running
job has been executed :-)
testjob's output:

removed crontab
<<<execution_status>>>
initiation_status="ok"
duration=132 termination_type=exited termination_id=0 corefile=no
cutime=8 cstime=6
<<<test_end>>>
<<<test_start>>>
tag=cron_deny01 stime=1509665854
cmdline="cron_deny01"
contacts=""
analysis=exit
<<<test_output>>>
This script contains bashism that needs to be fixed!
No passwd entry for user 'cd_user1'
No passwd entry for user 'cd_user2'
userdel: user 'cd_user1' does not exist
userdel: user 'cd_user2' does not exist
This script contains bashism that needs to be fixed!
TEST: /etc/cron.deny should allow only allow those who are not in the file to
run cron jobs.
(1) TEST THAT PERSON NOT IN /etc/cron.deny IS ABLE TO RUN JOB.
sleeping for 130 seconds....
Cron allowed user to execute test job, TEST PASSED
This script contains bashism that needs to be fixed!
(2) TEST THAT PERSON IN /etc/cron.deny IS NOT ABLE TO RUN JOB.
You (cd_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
Error while adding crontab for user cd_user2
sleeping for 130 seconds....
Cron did not allow user to execute job , TEST PASSED
You (cd_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
<<<execution_status>>>
initiation_status="ok"
duration=261 termination_type=exited termination_id=0 corefile=no
cutime=15 cstime=7
<<<test_end>>>
<<<test_start>>>
tag=cron_allow01 stime=1509666115
cmdline="cron_allow01"
contacts=""
analysis=exit
<<<test_output>>>
This script contains bashism that needs to be fixed!
No passwd entry for user 'ca_user1'
No passwd entry for user 'ca_user2'
userdel: user 'ca_user1' does not exist
userdel: user 'ca_user2' does not exist
This script contains bashism that needs to be fixed!
TEST: /etc/cron.allow should only allow those in the file to
run cron jobs.
(1) TEST THAT PERSON IN /etc/cron.allow IS ABLE TO RUN JOB.
backup crontab....
no crontab for ca_user1
sleeping for 130 seconds....
Cron allowed user to execute test job, TEST PASSED
restore old crontab...
This script contains bashism that needs to be fixed!
(2) TEST THAT PERSON NOT IN /etc/cron.allow IS NOT ABLE TO RUN JOB.
backup crontab....
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
Error while adding crontab for user ca_user2
sleeping for 130 seconds....
Cron did not allow user to execute job , TEST PASSED
restore old crontab...
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
<<<execution_status>>>
initiation_status="ok"
duration=261 termination_type=exited termination_id=0 corefile=no
cutime=12 cstime=8
<<<test_end>>>
<<<test_start>>>
tag=cron_dirs_checks01 stime=1509666376
cmdline="cron_dirs_checks01"
contacts=""
analysis=exit
<<<test_output>>>
incrementing stop
Checking /etc/cron.d
PASS: /etc/cron.d
Checking /etc/cron.daily
PASS: /etc/cron.daily
Checking /etc/cron.hourly
PASS: /etc/cron.hourly
Checking /etc/cron.monthly
PASS: /etc/cron.monthly
Checking /etc/cron.weekly
PASS: /etc/cron.weekly
Checking /var/spool/cron
PASS: /var/spool/cron
<<<execution_status>>>
initiation_status="ok"
duration=0 termination_type=exited termination_id=0 corefile=no
cutime=0 cstime=0
<<<test_end>>>
INFO: ltp-pan reported all tests PASS
LTP Version: 20170929-164-g09e19d4

       ###############################################################

            Done executing testcases.
            LTP Version:  20170929-164-g09e19d4
       ###############################################################

```

jstancek commented:
OK, so you're targeting on FILES section in man page. \b doesn't seem to be necessary now that there is [[:space:]]$. Some cron packages ship cron.deny, which is something we could look for too if man-page approach fails.

So I'm thinking, let's try the man page. If we fail, check for common paths, otherwise guess it will be /etc/cron.*. How does that sound?

```
setup_path() {
	CRON_ALLOW=$(man crontab 2>/dev/null | grep -m 1 -o '/[\/a-z.]*allow[[:space:]]*$')
	CRON_DENY=$(man crontab 2>/dev/null | grep -m 1 -o '/[\/a-z.]*deny[[:space:]]*$')

	if [ -n "$CRON_ALLOW" -a -n "$CRON_DENY" ]; then
		return
	fi

	if [ -e /var/spool/cron/deny ]; then
		export CRON_ALLOW=/var/spool/cron/allow
		export CRON_DENY=/var/spool/cron/deny
		return
	fi

	export CRON_ALLOW=/etc/cron.allow
	export CRON_DENY=/etc/cron.deny
}
```

lucasrangit commented:
> OK, so you're targeting on FILES section in man page. \b doesn't seem to be necessary now that there is [[:space:]]$.

@jstancek, true. In fact, after more testing across different distros, I find I can simplify it even further to just '$'.

> So I'm thinking, let's try the man page. If we fail, check for common paths, otherwise guess it will be /etc/cron.*. How does that sound?

Sure, that works. The path `/var/spool/cron/deny` is used in SLES 8.3, which includes the the vixie-cron man pages, is already matched by the grep. Is there a distro where this is not the case so I can test it? I don't think it's right to add code to LTP that I have not tested explicitly ;)

lucasrangit commented:
Tested on Debian 9 (stretch):

```
$ sudo /opt/ltp/runltp -f admin_tools -s cron
INFO: creating /opt/ltp/output directory
INFO: creating /opt/ltp/results directory
Checking for required user/group ids

'nobody' user id and group found.
'bin' user id and group found.
'daemon' user id and group found.
Users group found.
Sys group found.
Required users/groups exist.
If some fields are empty or look unusual you may have an old version.
Compare to the current minimal requirements in Documentation/Changes.
 
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
Linux magma.builder 4.9.0-4-amd64 #1 SMP Debian 4.9.51-1 (2017-09-28) x86_64 GNU/Linux
 
Gnu C                  gcc (Debian 6.3.0-18) 6.3.0 20170516
Gnu make               4.1
util-linux             linux 2.29.2
mount                  linux 2.29.2 (libmount 2.29.2: selinux, btrfs, assert, debug)
modutils               23
e2fsprogs              1.43.4
Linux C Library        > libc.2.24
Dynamic linker (ldd)   2.24
Procps                 3.3.12
iproute2               iproute2-ss161212
iputils                iputils-s20161105
Kbd                    2.0.3
Sh-utils               8.26
Modules Loaded         fuse coretemp crct10dif_pclmul crc32_pclmul ghash_clmulni_intel ppdev vmw_balloon sg intel_rapl_perf joydev evdev pcspkr serio_raw nfit libnvdimm vmwgfx ttm drm_kms_helper parport_pc shpchp parport battery ac button drm vmw_vsock_vmci_transport vsock vmw_vmci ip_tables x_tables autofs4 ext4 crc16 jbd2 crc32c_generic fscrypto ecb mbcache sd_mod sr_mod cdrom ata_generic crc32c_intel aesni_intel aes_x86_64 glue_helper lrw gf128mul ablk_helper cryptd psmouse e1000 mptspi scsi_transport_spi mptscsih mptbase ata_piix i2c_piix4 libata scsi_mod floppy

free reports:
              total        used        free      shared  buff/cache   available
Mem:        2052528       90796      245812        6676     1715920     1791676
Swap:       4193276           0     4193276

/proc/cpuinfo
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 94
model name	: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
stepping	: 3
microcode	: 0xba
cpu MHz		: 2711.140
cache size	: 8192 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 22
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm 3dnowprefetch epb fsgsbase smep dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bugs		:
bogomips	: 5424.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:

processor	: 1
vendor_id	: GenuineIntel
cpu family	: 6
model		: 94
model name	: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
stepping	: 3
microcode	: 0xba
cpu MHz		: 2711.140
cache size	: 8192 KB
physical id	: 2
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 2
initial apicid	: 2
fpu		: yes
fpu_exception	: yes
cpuid level	: 22
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm 3dnowprefetch epb fsgsbase smep dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bugs		:
bogomips	: 5424.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:

no big block device was specified on commandline.
Tests which require a big block device are disabled.
You can specify it with option -z
COMMAND:    /opt/ltp/bin/ltp-pan   -e -S   -a 31999     -n 31999 -p -f /tmp/ltp-PIa7Q7c33V/alltests -l /opt/ltp/results/LTP_RUN_ON-2017_11_03-10h_16m_23s.log  -C /opt/ltp/output/LTP_RUN_ON-2017_11_03-10h_16m_23s.failed -T /opt/ltp/output/LTP_RUN_ON-2017_11_03-10h_16m_23s.tconf
INFO: Restricted to cron
LOG File: /opt/ltp/results/LTP_RUN_ON-2017_11_03-10h_16m_23s.log
FAILED COMMAND File: /opt/ltp/output/LTP_RUN_ON-2017_11_03-10h_16m_23s.failed
TCONF COMMAND File: /opt/ltp/output/LTP_RUN_ON-2017_11_03-10h_16m_23s.tconf
Running tests.......
<<<test_start>>>
tag=cron02 stime=1509729384
cmdline="cron02"
contacts=""
analysis=exit
<<<test_output>>>
userdel: user 'c01_user' does not exist
Running as user c01_user...
new job added successfully
sleeping for 130 seconds...
Testjob running
job has been executed :-)
testjob's output:

removed crontab
<<<execution_status>>>
initiation_status="ok"
duration=132 termination_type=exited termination_id=0 corefile=no
cutime=9 cstime=4
<<<test_end>>>
<<<test_start>>>
tag=cron_deny01 stime=1509729516
cmdline="cron_deny01"
contacts=""
analysis=exit
<<<test_output>>>
This script contains bashism that needs to be fixed!
No passwd entry for user 'cd_user1'
No passwd entry for user 'cd_user2'
userdel: user 'cd_user1' does not exist
userdel: user 'cd_user2' does not exist
This script contains bashism that needs to be fixed!
TEST: /etc/cron.deny should allow only allow those who are not in the file to
run cron jobs.
(1) TEST THAT PERSON NOT IN /etc/cron.deny IS ABLE TO RUN JOB.
sleeping for 130 seconds....
Cron allowed user to execute test job, TEST PASSED
This script contains bashism that needs to be fixed!
(2) TEST THAT PERSON IN /etc/cron.deny IS NOT ABLE TO RUN JOB.
You (cd_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
Error while adding crontab for user cd_user2
sleeping for 130 seconds....
Cron did not allow user to execute job , TEST PASSED
You (cd_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
<<<execution_status>>>
initiation_status="ok"
duration=261 termination_type=exited termination_id=0 corefile=no
cutime=14 cstime=10
<<<test_end>>>
<<<test_start>>>
tag=cron_allow01 stime=1509729777
cmdline="cron_allow01"
contacts=""
analysis=exit
<<<test_output>>>
This script contains bashism that needs to be fixed!
No passwd entry for user 'ca_user1'
No passwd entry for user 'ca_user2'
userdel: user 'ca_user1' does not exist
userdel: user 'ca_user2' does not exist
This script contains bashism that needs to be fixed!
TEST: /etc/cron.allow should only allow those in the file to
run cron jobs.
(1) TEST THAT PERSON IN /etc/cron.allow IS ABLE TO RUN JOB.
backup crontab....
no crontab for ca_user1
sleeping for 130 seconds....
Cron allowed user to execute test job, TEST PASSED
restore old crontab...
This script contains bashism that needs to be fixed!
(2) TEST THAT PERSON NOT IN /etc/cron.allow IS NOT ABLE TO RUN JOB.
backup crontab....
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
Error while adding crontab for user ca_user2
sleeping for 130 seconds....
Cron did not allow user to execute job , TEST PASSED
restore old crontab...
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
<<<execution_status>>>
initiation_status="ok"
duration=262 termination_type=exited termination_id=0 corefile=no
cutime=13 cstime=8
<<<test_end>>>
<<<test_start>>>
tag=cron_dirs_checks01 stime=1509730039
cmdline="cron_dirs_checks01"
contacts=""
analysis=exit
<<<test_output>>>
incrementing stop
Checking /etc/cron.d
PASS: /etc/cron.d
Checking /etc/cron.daily
PASS: /etc/cron.daily
Checking /etc/cron.hourly
PASS: /etc/cron.hourly
Checking /etc/cron.monthly
PASS: /etc/cron.monthly
Checking /etc/cron.weekly
PASS: /etc/cron.weekly
Checking /var/spool/cron
PASS: /var/spool/cron
<<<execution_status>>>
initiation_status="ok"
duration=0 termination_type=exited termination_id=0 corefile=no
cutime=0 cstime=0
<<<test_end>>>
INFO: ltp-pan reported all tests PASS
LTP Version: 20170929-164-g09e19d4b3

       ###############################################################

            Done executing testcases.
            LTP Version:  20170929-164-g09e19d4b3
       ###############################################################
```

lucasrangit commented:
Tested on CentOS 7.4:

```
$ sudo /opt/ltp/runltp -f admin_tools -s cron
INFO: creating /opt/ltp/output directory
INFO: creating /opt/ltp/results directory
Checking for required user/group ids

'nobody' user id and group found.
'bin' user id and group found.
'daemon' user id and group found.
Users group found.
Sys group found.
Required users/groups exist.
If some fields are empty or look unusual you may have an old version.
Compare to the current minimal requirements in Documentation/Changes.
 
CentOS Linux release 7.4.1708 (Core) 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.4.1708 (Core) 
CentOS Linux release 7.4.1708 (Core) 
Linux localhost.localdomain 3.10.0-693.5.2.el7.x86_64 #1 SMP Fri Oct 20 20:32:50 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
 
Gnu C                  gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)
Gnu make               3.82
util-linux             linux 2.23.2
mount                  linux 2.23.2 (libmount 2.23.0: selinux, debug, assert)
modutils               20
e2fsprogs              1.42.9
Linux C Library        > libc.2.17
Dynamic linker (ldd)   2.17
Procps                 3.3.10
Net-tools              2.10-alpha
iproute2               iproute2-ss130716
iputils                iputils-s20160308
Kbd                    1.15.5
Sh-utils               8.22
Modules Loaded         vmhgfs vmw_vsock_vmci_transport vsock coretemp iosf_mbi crc32_pclmul ghash_clmulni_intel aesni_intel ppdev lrw gf128mul glue_helper ablk_helper cryptd vmw_balloon pcspkr joydev nfit sg libnvdimm vmw_vmci shpchp i2c_piix4 parport_pc parport nfsd auth_rpcgss nfs_acl lockd grace sunrpc ip_tables xfs libcrc32c sr_mod cdrom ata_generic pata_acpi sd_mod crc_t10dif crct10dif_generic vmwgfx drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops ttm drm ata_piix libata mptspi crct10dif_pclmul crct10dif_common scsi_transport_spi crc32c_intel serio_raw mptscsih e1000 mptbase i2c_core floppy dm_mirror dm_region_hash dm_log dm_mod

free reports:
              total        used        free      shared  buff/cache   available
Mem:        1016552      107448       64444        6904      844660      689856
Swap:       2097148           0     2097148

/proc/cpuinfo
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 94
model name	: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
stepping	: 3
microcode	: 0xba
cpu MHz		: 2712.000
cache size	: 8192 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 22
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm 3dnowprefetch epb fsgsbase smep dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bogomips	: 5424.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:

no big block device was specified on commandline.
Tests which require a big block device are disabled.
You can specify it with option -z
COMMAND:    /opt/ltp/bin/ltp-pan   -e -S   -a 1674     -n 1674 -p -f /tmp/ltp-z3OPR0PPmw/alltests -l /opt/ltp/results/LTP_RUN_ON-2017_11_03-23h_04m_28s.log  -C /opt/ltp/output/LTP_RUN_ON-2017_11_03-23h_04m_28s.failed -T /opt/ltp/output/LTP_RUN_ON-2017_11_03-23h_04m_28s.tconf
INFO: Restricted to cron
LOG File: /opt/ltp/results/LTP_RUN_ON-2017_11_03-23h_04m_28s.log
FAILED COMMAND File: /opt/ltp/output/LTP_RUN_ON-2017_11_03-23h_04m_28s.failed
TCONF COMMAND File: /opt/ltp/output/LTP_RUN_ON-2017_11_03-23h_04m_28s.tconf
Running tests.......
<<<test_start>>>
tag=cron02 stime=1509750268
cmdline="cron02"
contacts=""
analysis=exit
<<<test_output>>>
userdel: user 'c01_user' does not exist
Running as user c01_user...
new job added successfully
sleeping for 130 seconds...
Testjob running
job has been executed :-)
testjob's output:

removed crontab
<<<execution_status>>>
initiation_status="ok"
duration=132 termination_type=exited termination_id=0 corefile=no
cutime=4 cstime=14
<<<test_end>>>
<<<test_start>>>
tag=cron_deny01 stime=1509750400
cmdline="cron_deny01"
contacts=""
analysis=exit
<<<test_output>>>
This script contains bashism that needs to be fixed!
su: user cd_user1 does not exist
su: user cd_user2 does not exist
userdel: user 'cd_user1' does not exist
userdel: user 'cd_user2' does not exist
This script contains bashism that needs to be fixed!
TEST: /etc/cron.deny should allow only allow those who are not in the file to
run cron jobs.
(1) TEST THAT PERSON NOT IN /etc/cron.deny IS ABLE TO RUN JOB.
sleeping for 130 seconds....
Cron allowed user to execute test job, TEST PASSED
This script contains bashism that needs to be fixed!
(2) TEST THAT PERSON IN /etc/cron.deny IS NOT ABLE TO RUN JOB.
You (cd_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
Error while adding crontab for user cd_user2
sleeping for 130 seconds....
Cron did not allow user to execute job , TEST PASSED
You (cd_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
<<<execution_status>>>
initiation_status="ok"
duration=261 termination_type=exited termination_id=0 corefile=no
cutime=8 cstime=16
<<<test_end>>>
<<<test_start>>>
tag=cron_allow01 stime=1509750661
cmdline="cron_allow01"
contacts=""
analysis=exit
<<<test_output>>>
This script contains bashism that needs to be fixed!
su: user ca_user1 does not exist
su: user ca_user2 does not exist
userdel: user 'ca_user1' does not exist
userdel: user 'ca_user2' does not exist
This script contains bashism that needs to be fixed!
TEST: /etc/cron.allow should only allow those in the file to
run cron jobs.
(1) TEST THAT PERSON IN /etc/cron.allow IS ABLE TO RUN JOB.
backup crontab....
no crontab for ca_user1
sleeping for 130 seconds....
Cron allowed user to execute test job, TEST PASSED
restore old crontab...
This script contains bashism that needs to be fixed!
(2) TEST THAT PERSON NOT IN /etc/cron.allow IS NOT ABLE TO RUN JOB.
backup crontab....
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
Error while adding crontab for user ca_user2
sleeping for 130 seconds....
Cron did not allow user to execute job , TEST PASSED
restore old crontab...
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
<<<execution_status>>>
initiation_status="ok"
duration=261 termination_type=exited termination_id=0 corefile=no
cutime=5 cstime=19
<<<test_end>>>
<<<test_start>>>
tag=cron_dirs_checks01 stime=1509750922
cmdline="cron_dirs_checks01"
contacts=""
analysis=exit
<<<test_output>>>
incrementing stop
Checking /etc/cron.d
PASS: /etc/cron.d
Checking /etc/cron.daily
PASS: /etc/cron.daily
Checking /etc/cron.hourly
PASS: /etc/cron.hourly
Checking /etc/cron.monthly
PASS: /etc/cron.monthly
Checking /etc/cron.weekly
PASS: /etc/cron.weekly
Checking /var/spool/cron
PASS: /var/spool/cron
<<<execution_status>>>
initiation_status="ok"
duration=0 termination_type=exited termination_id=0 corefile=no
cutime=1 cstime=0
<<<test_end>>>
INFO: ltp-pan reported all tests PASS
LTP Version: 20170929-164-gb2ebb7c

       ###############################################################

            Done executing testcases.
            LTP Version:  20170929-164-gb2ebb7c
       ###############################################################

```

lucasrangit commented:
Fedora 26 test:

```
$ sudo /opt/ltp/runltp -f admin_tools -s cron
INFO: creating /opt/ltp/output directory
INFO: creating /opt/ltp/results directory
Checking for required user/group ids

'nobody' user id and group found.
'bin' user id and group found.
'daemon' user id and group found.
Users group found.
Sys group found.
Required users/groups exist.
If some fields are empty or look unusual you may have an old version.
Compare to the current minimal requirements in Documentation/Changes.
 
Fedora release 26 (Twenty Six)
NAME=Fedora
VERSION="26 (Twenty Six)"
ID=fedora
VERSION_ID=26
PRETTY_NAME="Fedora 26 (Twenty Six)"
ANSI_COLOR="0;34"
CPE_NAME="cpe:/o:fedoraproject:fedora:26"
HOME_URL="https://fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=26
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=26
PRIVACY_POLICY_URL=https://fedoraproject.org/wiki/Legal:PrivacyPolicy
Fedora release 26 (Twenty Six)
Fedora release 26 (Twenty Six)
Linux localhost.localdomain 4.11.8-300.fc26.x86_64 #1 SMP Thu Jun 29 20:09:48 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
 
Gnu C                  gcc (GCC) 7.2.1 20170915 (Red Hat 7.2.1-2)
Gnu make               4.2.1
util-linux             linux 2.29.1
mount                  linux 2.29.1 (libmount 2.29.1: selinux, btrfs, assert, debug)
modutils               24
e2fsprogs              1.43.4
Linux C Library        > libc.2.25
Dynamic linker (ldd)   2.25
Procps                 3.3.10
Net-tools              2.10-alpha
iproute2               iproute2-ss170501
iputils                iputils-s20161105
Kbd                    2.0.4
Sh-utils               8.27
Modules Loaded         fuse vmwgfx coretemp crct10dif_pclmul crc32_pclmul crc32c_intel ghash_clmulni_intel ttm ppdev drm_kms_helper vmw_balloon intel_rapl_perf drm joydev e1000 i2c_piix4 shpchp nfit vmw_vsock_vmci_transport vsock vmw_vmci tpm_tis parport_pc tpm_tis_core parport sunrpc tpm serio_raw mptspi scsi_transport_spi mptscsih mptbase ata_generic pata_acpi

free reports:
              total        used        free      shared  buff/cache   available
Mem:        1015420       77368       67372         472      870680      751980
Swap:       2097148         776     2096372

/proc/cpuinfo
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 94
model name	: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
stepping	: 3
microcode	: 0xba
cpu MHz		: 2712.000
cache size	: 8192 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 22
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc cpuid aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm 3dnowprefetch epb fsgsbase smep dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bugs		:
bogomips	: 5424.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:

no big block device was specified on commandline.
Tests which require a big block device are disabled.
You can specify it with option -z
COMMAND:    /opt/ltp/bin/ltp-pan   -e -S   -a 6356     -n 6356 -p -f /tmp/ltp-bUovq4Tncd/alltests -l /opt/ltp/results/LTP_RUN_ON-2017_11_04-00h_05m_01s.log  -C /opt/ltp/output/LTP_RUN_ON-2017_11_04-00h_05m_01s.failed -T /opt/ltp/output/LTP_RUN_ON-2017_11_04-00h_05m_01s.tconf
INFO: Restricted to cron
LOG File: /opt/ltp/results/LTP_RUN_ON-2017_11_04-00h_05m_01s.log
FAILED COMMAND File: /opt/ltp/output/LTP_RUN_ON-2017_11_04-00h_05m_01s.failed
TCONF COMMAND File: /opt/ltp/output/LTP_RUN_ON-2017_11_04-00h_05m_01s.tconf
Running tests.......
<<<test_start>>>
tag=cron02 stime=1509753902
cmdline="cron02"
contacts=""
analysis=exit
<<<test_output>>>
userdel: user 'c01_user' does not exist
Running as user c01_user...
new job added successfully
sleeping for 130 seconds...
Testjob running
job has been executed :-)
testjob's output:

removed crontab
<<<execution_status>>>
initiation_status="ok"
duration=131 termination_type=exited termination_id=0 corefile=no
cutime=6 cstime=12
<<<test_end>>>
<<<test_start>>>
tag=cron_deny01 stime=1509754033
cmdline="cron_deny01"
contacts=""
analysis=exit
<<<test_output>>>
This script contains bashism that needs to be fixed!
su: user cd_user1 does not exist
su: user cd_user2 does not exist
userdel: user 'cd_user1' does not exist
userdel: user 'cd_user2' does not exist
This script contains bashism that needs to be fixed!
TEST: /etc/cron.deny should allow only allow those who are not in the file to
run cron jobs.
(1) TEST THAT PERSON NOT IN /etc/cron.deny IS ABLE TO RUN JOB.
sleeping for 130 seconds....
Cron allowed user to execute test job, TEST PASSED
This script contains bashism that needs to be fixed!
(2) TEST THAT PERSON IN /etc/cron.deny IS NOT ABLE TO RUN JOB.
You (cd_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
Error while adding crontab for user cd_user2
sleeping for 130 seconds....
Cron did not allow user to execute job , TEST PASSED
You (cd_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
<<<execution_status>>>
initiation_status="ok"
duration=262 termination_type=exited termination_id=0 corefile=no
cutime=8 cstime=15
<<<test_end>>>
<<<test_start>>>
tag=cron_allow01 stime=1509754295
cmdline="cron_allow01"
contacts=""
analysis=exit
<<<test_output>>>
This script contains bashism that needs to be fixed!
su: user ca_user1 does not exist
su: user ca_user2 does not exist
userdel: user 'ca_user1' does not exist
userdel: user 'ca_user2' does not exist
This script contains bashism that needs to be fixed!
TEST: /etc/cron.allow should only allow those in the file to
run cron jobs.
(1) TEST THAT PERSON IN /etc/cron.allow IS ABLE TO RUN JOB.
backup crontab....
no crontab for ca_user1
sleeping for 130 seconds....
Cron allowed user to execute test job, TEST PASSED
restore old crontab...
This script contains bashism that needs to be fixed!
(2) TEST THAT PERSON NOT IN /etc/cron.allow IS NOT ABLE TO RUN JOB.
backup crontab....
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
Error while adding crontab for user ca_user2
sleeping for 130 seconds....
Cron did not allow user to execute job , TEST PASSED
restore old crontab...
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
You (ca_user2) are not allowed to use this program (crontab)
See crontab(1) for more information
<<<execution_status>>>
initiation_status="ok"
duration=261 termination_type=exited termination_id=0 corefile=no
cutime=8 cstime=16
<<<test_end>>>
<<<test_start>>>
tag=cron_dirs_checks01 stime=1509754556
cmdline="cron_dirs_checks01"
contacts=""
analysis=exit
<<<test_output>>>
incrementing stop
Checking /etc/cron.d
PASS: /etc/cron.d
Checking /etc/cron.daily
PASS: /etc/cron.daily
Checking /etc/cron.hourly
PASS: /etc/cron.hourly
Checking /etc/cron.monthly
PASS: /etc/cron.monthly
Checking /etc/cron.weekly
PASS: /etc/cron.weekly
Checking /var/spool/cron
PASS: /var/spool/cron
<<<execution_status>>>
initiation_status="ok"
duration=0 termination_type=exited termination_id=0 corefile=no
cutime=1 cstime=0
<<<test_end>>>
INFO: ltp-pan reported all tests PASS
LTP Version: 20170929-164-gb2ebb7c3e

       ###############################################################

            Done executing testcases.
            LTP Version:  20170929-164-gb2ebb7c3e
       ###############################################################

```

lucasrangit commented:
@danrue Thanks for sending this to the mailing list and adding your "Tested-by". I hope it will get merged soon.

metan-ucw commented:
The cron tests have been removed in 0fc9b8624bea, hence closing, thanks for the PR anyway.
